### PR TITLE
kind: Use quay registry image

### PR DESCRIPTION
**Pull request by oshoval**
Overcome docker.io rate limits

Signed-off-by: Or Shoval <oshoval@redhat.com>

### CODE REVIEW

**Comment by oshoval** [comment]
some tests are failing
but not due to this PR

/test check-up-kind-1.22-sriov

File: cluster-up/cluster/kind/common.sh [hunk 1]
```diff
@@ -94,7 +94,7 @@ function _run_registry() {
         docker rm $REGISTRY_NAME || true
         sleep 5
     done
-    docker run -d --network=${network} -p $HOST_PORT:5000  --restart=always --name $REGISTRY_NAME registry:2
+    docker run -d --network=${network} -p $HOST_PORT:5000  --restart=always --name $REGISTRY_NAME quay.io/libpod/registry:2.7
```
Comment (ormergi) [inline cluster-up/cluster/kind/common.sh:97]: I suggest to use version 2.6, it has higher security score:
https://quay.io/repository/libpod/registry?tab=tags

**Review by ormergi — COMMENTED** [review]
I couldn't find a the repository that publish this container (libpod org),
do you know if there is one? is it maintained?
If not I suggest to use mirror of the one that is currently used.

**Comment by oshoval** [comment]
> I couldn't find a the repository that publish this container (libpod org), do you know if there is one? is it maintained? If not I suggest to use mirror of the one that is currently used.

It is the same one that we are using on VM based providers in kubevirtci
so i think it fine to use it, 
don't know beside that about libpod

File: cluster-up/cluster/kind/common.sh (hunk 1 above)
Comment (oshoval) [reply cluster-up/cluster/kind/common.sh:97]: using the same version we are using on VM based providers, i think it is better to have the same version

**Comment by oshoval** [comment]
Thanks all

/retest

**Comment by zhlhahaha** [comment]
hi @oshoval 
This registry image does not support multi-arch. What can I do to make it support arm64?

**Comment by oshoval** [comment]
> hi @oshoval This registry image does not support multi-arch. What can I do to make it support arm64?

Hi @zhlhahaha 
we can either expose env var for it, and then manually export it when needed,
or to mirror the original image to quay.io/kubevirtci and use it, there is a script that mirror images to quay, i need to recall where it is

**Comment by oshoval** [comment]
@zhlhahaha can you check this image please ? `quay.io/kubevirtci/library-registry:2.7.1`
if it is good i will create a PR for it, it should be a mirror of `library/registry:2.7.1`

**Comment by zhlhahaha** [comment]
> quay.io/kubevirtci/library-registry:2.7.1

I am unable to pull this image.
```
Unable to find image 'quay.io/kubevirtci/library-registry:2.7.1' locally
docker: Error response from daemon: unauthorized: access to the requested resource is not authorized.
```
It seems that I need to login quay.io, then I can pull this image.

**Comment by oshoval** [comment]
> It seems that I need to login quay.io, then I can pull this image.

I will ask to make it public

**Comment by oshoval** [comment]
@zhlhahaha 
can you please retry ?

**Comment by oshoval** [comment]
https://hub.docker.com/layers/registry/library/registry/2.7.1/images/sha256-e9c2d646078211f8353e482caac9d9b12f38b42bc29cb30c453841cf144037b1?context=explore

This one works fine ? (it mentions arm64)
If so we might need to improve the script that mirror images to support sha and then to mirror it
However, I am not sure i will be able to work on it atm

**Comment by zhlhahaha** [comment]
> https://hub.docker.com/layers/registry/library/registry/2.7.1/images/sha256-e9c2d646078211f8353e482caac9d9b12f38b42bc29cb30c453841cf144037b1?context=explore
> 
> This one works fine ? (it mentions arm64) If so we might need to improve the script that mirror images to support sha (unless it already does) and then to mirror it However, I am not sure i will be able to work on it atm

Yes, I tried `docker pull registry:2.7.1`, the registry image works fine on Arm64. Can I do some thing on it?

> created this meanwhile, in case it will be merged we can reuse the previous image for now (but on quay) [kubevirt/project-infra#2030](https://github.com/kubevirt/project-infra/pull/2030)

Thanks!

**Comment by zhlhahaha** [comment]
When I use `docker inspect quay.io/kubevirtci/library-registry:2.7.1`

```
[
    {
        "Id": "sha256:b8604a3fe8543c9e6afc29550de05b36cd162a97aa9b2833864ea8a5be11f3e2",
        "RepoTags": [
            "quay.io/kubevirtci/library-registry:2.7.1"
        ],
        "RepoDigests": [
            "quay.io/kubevirtci/library-registry@sha256:36cb5b157911061fb610d8884dc09e0b0300a767a350563cbfd88b4b85324ce4"
        ],
```

**Comment by oshoval** [comment]
I believe it should be docker pull and push instead of skopeo that currently copies the images then

**Comment by oshoval** [comment]
Hi @zhlhahaha 
Can you check please if this image supports arm64 ?
`quay.io/oshoval/registry:2.7.1`
It is `library/registry:2.7.1` which I docker pull and push instead of using skopeo copy

**Comment by zhlhahaha** [comment]
> Can you check please if this image supports arm64 ?

I tried, it is still only for amd64. I tried `docker inspect quay.io/oshoval/registry:2.7.1` on Arm64 server and it shows `        "Architecture": "amd64",`

**Comment by zhlhahaha** [comment]
Maybe I can give it a try how to use docker pull and docker push to get multi-arch image.

**Comment by oshoval** [comment]
> Maybe I can give it a try how to use docker pull and docker push to get multi-arch image.

Thanks for checking,
Would be awesome
If needed I am available at slack / email, slack has some quay support as well

**Comment by zhlhahaha** [comment]
> Thanks for checking, Would be awesome If needed I am available at slack / email, slack has some quay support as well

No problem

**Comment by zhlhahaha** [comment]
Can you give this script a try? Change the `targetRepo` to your repository name.
```
#!/bin/bash -x
archs="arm64 x86_64"
image="registry"
tag="2.7.1"
targetRepo="zhlhahaha"
manifestList=""

for arch in ${archs}; do
    docker pull --platform linux/"${arch}" "${image}":"${tag}"
    docker tag "${image}":"${tag}" "${targetRepo}"/"${image}":"${tag}"-"${arch}"
    docker push "${targetRepo}"/"${image}":"${tag}"-"${arch}"
    docker rmi "${image}":"${tag}"
    manifestList="${manifestList} --amend ""${targetRepo}"/"${image}":"${tag}"-"${arch}"
done

docker manifest create "${targetRepo}"/"${image}":"${tag}" ${manifestList}
docker manifest push "${targetRepo}"/"${image}":"${tag}"
```

**Comment by oshoval** [comment]
> Can you give this script a try? Change the `targetRepo` to your repository name.
> 
> ```
> #!/bin/bash -x
> archs="arm64 x86_64"
> image="registry"
> tag="2.7.1"
> targetRepo="zhlhahaha"
> manifestList=""
> 
> for arch in ${archs}; do
>     docker pull --platform linux/"${arch}" "${image}":"${tag}"
>     docker tag "${image}":"${tag}" "${targetRepo}"/"${image}":"${tag}"-"${arch}"
>     docker push "${targetRepo}"/"${image}":"${tag}"-"${arch}"
>     docker rmi "${image}":"${tag}"
>     manifestList="${manifestList} --amend ""${targetRepo}"/"${image}":"${tag}"-"${arch}"
> done
> 
> docker manifest create "${targetRepo}"/"${image}":"${tag}" ${manifestList}
> docker manifest push "${targetRepo}"/"${image}":"${tag}"
> ```

Very nice thanks
pushed this one `quay.io/oshoval/registry:2.7.1` again now with the 2 manifests, can you please try ?

**Comment by zhlhahaha** [comment]
> quay.io/oshoval/registry:2.7.1

I tried it on both x86_64 server and arm64 server, both works well. Thanks.

**Comment by oshoval** [comment]
> > quay.io/oshoval/registry:2.7.1
> 
> I tried it on both x86_64 server and arm64 server, both works well. Thanks.

Thank you ;] will update our script to adapt the method

**Comment by oshoval** [comment]
btw maybe buildx is the official way ?
https://www.acervera.com/blog/2020/04/sbt-docker-buildx-multi-arch/

**Comment by zhlhahaha** [comment]
> btw maybe buildx is the official way ? https://www.acervera.com/blog/2020/04/sbt-docker-buildx-multi-arch/ (interesting part on the bottom)
> 
> For example https://github.com/nmstate/kubernetes-nmstate/blob/main/hack/build-push-container.docker.sh unless our scenario is little different

`docker buildx` include two steps:
1. build multi-arch image
2. generate multi-arch manifest and push
Our scenario is a little bit different, as we do not have the build step.

**Comment by oshoval** [comment]
FYI 
https://github.com/kubevirt/project-infra/pull/2061

**Pull request merged**